### ISSUE TITLE
Dealing with numerical issues (Cholesky decomposition)

### ISSUE THREAD

**Issue by martiningram** [tensorflow/probability#195]
Hi everyone,

Thanks for putting this great package together!

I am writing Gaussian Process models in tfp and sampling the kernel hyperparameters using HMC. Sometimes, due to numerical issues, the resulting kernel is no longer positive definite, causing the Cholesky decomposition to fail and raising an Exception. I would like to reject that proposal without crashing everything.

To catch the Exception, I tried to use eager mode. I thought I could do the following:

```
def compute_cholesky_if_possible(x):

    try:
        cholesky = tf.cholesky(x)
        return cholesky
    except Exception:
        print('Matrix not positive-definite!')
        return x * tf.constant(np.nan)
```

That half-works, but eventually crashes too, in the gradient of the Cholesky decomposition.

Is there a good way of handling this?

Thanks!

**Comment by csuter** [comment]
Hi Martin, are you using TransformedTransitionKernel to ensure the HMC sampler stays in the region of allowable values? You might need to Chain a tiny shift-only Affine bijector in addition to a positivity constraint (eg Exp or Softplus).

**Comment by martiningram** [comment]
Hi Christopher,

Thanks for the quick response! No, I am just using `tfp.mcmc.HamiltonianMonteCarlo` with `step_size_update_fn=tfp.mcmc.make_simple_step_size_update_policy()` for my parameters, and for those constrained to be positive, I pass them to the `log_posterior` function unconstrained, and then constrain them with a `tf.exp` as well as a log jacobian correction. The kernel matrix I am trying to Cholesky should be positive definite by construction, but I guess it isn't for some input values, e.g. very large parameters (or very small).

In Stan, the sampler often produces messages like:

> Informational Message: The current Metropolis proposal is about to be rejected because of the following issue:
> Exception: cholesky_decompose: Matrix m is not positive definite  (in 'unknown file name' at line 62)
> If this warning occurs sporadically, such as for highly constrained variable types like covariance matrices, then the sampler is fine,
> but if this warning occurs often then your model may be either severely ill-conditioned or misspecified.

at the start of sampling, and sometimes during, which I think is the same issue as here. Hope that clarifies things. But I'll look into the TransformedTransitionKernel!

**Comment by csuter** [comment]
That is a much nicer error message than we are able to conjure presently :)

You can also turn on the validate_args flag on your kernel instance and it might catch numerically zeroed values. This will help diagnose but won't solve it. There's an example in the docstrings for the GP regression model distribution on how to use TxTxKernel (sorry for brevity; on phone) but i don't think it includes the Affine shift which will help when the values get too near zero. Let us know if it's not clear how to adapt that example or if you run into other problems.

**Comment by mathDR** [comment]
The typical solution to this problem is to add "jitter" to the covariance to make it more numerically stable.  A possible solution might look like:
```
def compute_cholesky_if_possible(x):
    try:
        cholesky = tf.cholesky(x)
        return cholesky
    except Exception:
        print('Matrix not positive-definite! Adding Jitter')
        jitter = 1e-9
        while jitter < 1.0:
            try:
                cholesky = tf.cholesky(x + tf.Diagonal(jitter))
                return cholesky
            except Exception:
                jitter = jitter * 10
        return x * tf.constant(np.nan)
```

where `tf.Diagonal(jitter)` creates a diagonal matrix of same size as `x` with `jitter` along the diagonal.

**Comment by csuter** [comment]
Thanks, @mathDR, agreed. We have a `jitter` parameter on the GP Distribution classes for this purpose. However, when exploring the model parameter space with (unconstrained) HMC, it's quite easy to end up regions of parameter space that lead to non-PD kernel matrices. We actually have a default jitter of 1e-6, so if @martiningram was hitting bad Cholesky's (and hadn't overriden the jitter), it was almost certainly due to getting non-sensical kernel params from the HMC exploration.

**Comment by mathDR** [comment]
Ah, okay sorry I misunderstood the question.  Having said that, how should this fail?  You could always find the closest PD kernel to the current kernel, but as you noted, you shouldn't be in this area of the parameter space anyway...

**Comment by csuter** [comment]
> Having said that, how should this fail?
Not sure what you're referring to here -- the jitter search, or the HMC (as written above), or something else?

The HMC procedure can fail, in my experience, when the unconstrained values go very negative and the constrained values (exp'ed, say), become numerically too close to zero. Hence my suggestion of combining the `Exp` `Bijector` (actually `Softplus` is often better than `Exp`), with an `Affine(shift=very_small_value, scale=identity)`. This keeps the kernel params (amplitude and length_scale, say) sufficiently far from zero to avoid numerical issues.

**Comment by mathDR** [comment]
Right.  I meant fail in the area of the HMC calculations.  If you are in an area where the parameter values are leading to nonsensical Kernel matrices, should the algorithm fail, or should there be a way to "get back" to good areas of the search space?

If the latter, then wouldn't oscillations be a problem (in some instances)?

**Comment by csuter** [comment]
I see. The programming model of TensorFlow is such that you really want to avoid the kind of conditional behavior out control flow that would be necessary to "special case" bad regions of model parameter space. Instead it's best to design the algorithm to never be able to get into those regions in the first place, which is what the TransformedTransitionKernel is there to do. Of course, this is probably good practice irrespective of the programming model (declarative graph building or imperative "define by run")
